src/MongoDB.Driver.Core/Core/Compression/Zstd/ZstandardStream.cs

```diff
@@ -0,0 +1,380 @@
+        {
+            if (!CanRead) throw new InvalidDataException("Read is not accessible.");
+
+            var uncompressedOutputBuffer = new BufferInfo(outputBytes, outputOffset);
+
+            using (var operationContext = _nativeWrapper.InitializeOperationContext(
+                _streamReadHelper.CompressedBufferInfo,
+                uncompressedOutputBuffer)) // operation result
+            {
+                var length = 0; // the result size
+
+                while (count > 0)
+                {
+                    var currentDataPosition = _streamReadHelper.ReadCompressedStreamToBufferIfAvailable(
+                        recommendedCompressedSize: _nativeWrapper.RecommendedInputSize,
+                        out int remainingCompressedBufferSize);
+
+                    // decompress input to output
+                    _nativeWrapper.Decompress(
+                        operationContext,
+                        compressedOffset: currentDataPosition,
+                        compressedSize: remainingCompressedBufferSize,
+                        uncompressedSize: count,
+                        out int compressedBufferPosition,
+                        out int uncompressedBufferPosition);
+
+                    if (!_streamReadHelper.TryPrepareDataForNextAttempt(
+                        compressedBufferPosition,
+                        uncompressedBufferPosition))
+                    {
+                        break;
+                    }
+
+                    length += uncompressedBufferPosition;
+                    count -= uncompressedBufferPosition;
+                }
+
+                return length;
+            }
+        }
+
+        public override void Write(byte[] inputBytes, int inputOffset, int count) // Compress
+        {
+            if (!CanWrite) throw new InvalidDataException("Write is not accessible.");
+
+            var uncompressedInputInfo = new BufferInfo(inputBytes, inputOffset);
+
+            using (var operationContext = _nativeWrapper.InitializeOperationContext(
+                _streamWriteHelper.CompressedBufferInfo, // operation result
+                uncompressedInputInfo))
+            {
+                while (count > 0)
+                {
+                    var currentAttemptSize = Math.Min(count, _nativeWrapper.RecommendedInputSize);
+
+                    // compress input to output
+                    _nativeWrapper.Compress(
+                        operationContext,
+                        compressedSize: _nativeWrapper.RecommendedOutputSize,
+                        uncompressedSize: currentAttemptSize,
+                        out var compressedBufferPosition,
+                        out var uncompressedBufferPosition);
+
+                    _streamWriteHelper.WriteBufferToCompressedStream(count: compressedBufferPosition);
+
+                    // calculate progress in input buffer
+                    count -= uncompressedBufferPosition;
```

Comment:
Counts and positions are not the same thing. Are you sure this is correct?

Reply:
I've added a new variable `remainingCount`, I hope it will be more clear.